### PR TITLE
Add edit form/bulk edit form support for Notes.

### PR DESCRIPTION
**Pull request by smk4664**
# Closes: #2105 
# What's Changed

This PR adds in support for Primary Model and Organizational Models using the NautobotEditForm or NautobotBulkEditForm to add Notes during the creation/edit of an object.

image
image

# TODO

- [x] Explanation of Change(s)
- [x] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [x] Documentation Updates (when adding/changing features)
- [x] Example Plugin Updates (when adding/changing features)
- [x] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/extras/forms/mixins.py [hunk 1]
```diff
@@ -137,6 +140,49 @@ def __init__(self, *args, **kwargs):
...
+class NoteModelFormMixin(forms.ModelForm):
+    def __init__(self, *args, **kwargs):
+        self.obj_type = ContentType.objects.get_for_model(self._meta.model)
+
+        super().__init__(*args, **kwargs)
+
+        self.fields["object_note"] = CommentField()
+        self.fields["object_note"].label = "Note"
+
+    def save_note(self, *, instance, user):
```
Comment (glennmatthews) [inline nautobot/extras/forms/mixins.py:174]: Any chance we can consolidate these two form classes? They look like they're nearly identical.

File: nautobot/extras/tests/test_forms.py [hunk 2]
```diff
@@ -131,6 +136,68 @@ def test_create_job_hooks_with_same_job_common_action_same_content_type(self):
...
+
+        form = SiteForm(data=dict(**self.site_form_base_data, **{"object_note": "This is a test."}))
+        self.assertTrue(form.is_valid())
+        obj = form.save()
+        form.save_note(
+            instance=obj,
+            user=self.user,
+        )
+        self.assertEqual(1, Note.objects.count())
+        self.assertEqual("This is a test.", Note.objects.first().note)
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_forms.py:165]: Perhaps assert that this Note is linked to the correct object instance and user as well?

File: nautobot/extras/forms/mixins.py (hunk 1 above)
Comment (smk4664) [reply nautobot/extras/forms/mixins.py:174]: Same with RelationshipModel... The differences are the trick here.  They have different base forms (tested) and different methods for accessing the model.

File: nautobot/extras/tests/test_forms.py (hunk 2 above)
Comment (smk4664) [reply nautobot/extras/tests/test_forms.py:165]: Sure, I will add that test in.

File: nautobot/extras/forms/mixins.py (hunk 1 above)
Comment (smk4664) [reply nautobot/extras/forms/mixins.py:174]: I think I got it though, running tests.

File: nautobot/core/views/generic.py [hunk 3]
```diff
@@ -966,7 +969,9 @@ def post(self, request, **kwargs):
                 form_custom_fields = getattr(form, "custom_fields", [])
                 form_relationships = getattr(form, "relationships", [])
                 standard_fields = [
-                    field for field in form.fields if field not in form_custom_fields + form_relationships + ["pk"]
+                    field
+                    for field in form.fields
+                    if field not in form_custom_fields + form_relationships + ["pk"] + ["object_note"]
```
Comment (glennmatthews) [inline nautobot/core/views/generic.py:974]: tagging @HanlinMiao - we'll want to make sure we get equivalent updates into the NautobotViewSet work.

File: nautobot/extras/forms/mixins.py [hunk 4]
```diff
@@ -137,6 +140,44 @@ def __init__(self, *args, **kwargs):
...
 
+class NoteFormBase:
+    """Base fore the NoteModelFormMixin and NoteModelBulkEditFormMixin."""
+
+    def append_field(self):
+        self.fields["object_note"] = CommentField()
+        self.fields["object_note"].label = "Note"
+
+    def save_note(self, *, instance, user):
+        if "object_note" in self.cleaned_data:
```
Comment (glennmatthews) [inline nautobot/extras/forms/mixins.py:151]: Just checking - if no note text is specified (the field is left blank), we don't create an empty Note, do we?

File: nautobot/extras/forms/base.py [hunk 5]
```diff
@@ -21,10 +23,10 @@
 #
 
 
-class NautobotModelForm(BootstrapMixin, CustomFieldModelForm, RelationshipModelFormMixin):
+class NautobotModelForm(BootstrapMixin, CustomFieldModelForm, RelationshipModelFormMixin, NoteModelFormMixin):
     """
     This class exists to combine common functionality and is used to inherit from throughout the
-    codebase where all three of BootstrapMixin, CustomFieldModelForm and RelationshipModelForm are
+    codebase where all four of BootstrapMixin, CustomFieldModelForm, RelationshipModelForm and NoteModelForm are
     needed.
```
Comment (glennmatthews) [inline nautobot/extras/forms/base.py:30]:
```suggestion
    codebase where all of BootstrapMixin, CustomFieldModelForm, RelationshipModelForm and NoteModelForm are
    needed.
```

File: nautobot/extras/forms/mixins.py [hunk 6]
```diff
@@ -137,6 +140,44 @@ def __init__(self, *args, **kwargs):
             self.custom_fields.append(field_name)
 
 
+class NoteFormBase:
+    """Base fore the NoteModelFormMixin and NoteModelBulkEditFormMixin."""
+
+    def append_field(self):
+        self.fields["object_note"] = CommentField()
+        self.fields["object_note"].label = "Note"
```
Comment (glennmatthews) [inline nautobot/extras/forms/mixins.py:148]: Do we need this custom approach? Or can we just have:

```python
class NoteFormBase:
    """Base for the NoteModelFormMixin and NoteModelBulkEditFormMixin."""

    object_note = CommentField(label="Note")
```

File: nautobot/extras/tests/test_forms.py [hunk 7]
```diff
@@ -131,6 +136,71 @@ def test_create_job_hooks_with_same_job_common_action_same_content_type(self):
...
 
+class NoteModelFormTestCase(TestCase):
+    """
+    TestNoteModelForm validation and saving.
+    """
+
+    @classmethod
+    def setUpTestData(cls):
+        active = Status.objects.get(slug="active")
+        cls.user = User.objects.create(username="formmuser1")
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_forms.py:147]:
```suggestion
        cls.user = User.objects.create(username="formuser1")
```

File: nautobot/extras/forms/mixins.py (hunk 4 above)
Comment (smk4664) [reply nautobot/extras/forms/mixins.py]: Good catch, added additional to the conditional to ensure this doesn't happen.

File: nautobot/extras/forms/mixins.py (hunk 6 above)
Comment (smk4664) [reply nautobot/extras/forms/mixins.py]: Yes, but I had to add forms.Form as the base class.

File: nautobot/docs/development/style-guide.md [hunk 8]
```diff
@@ -63,7 +63,7 @@ New dependencies can be added to the project via the `poetry add` command. This
 
 * Similarly, for filter forms, `nautobot.extras.forms.NautobotFilterForm` combines `nautobot.utilities.forms.BootstrapMixin`, `nautobot.extras.forms.CustomFieldModelFilterFormMixin`, and `nautobot.extras.forms.RelationshipModelFilterFormMixin`, and should be used where appropriate.
 
-* Similarly, for bulk-edit forms, `nautobot.extras.forms.NautobotBulkEditForm` combines `nautobot.utilities.forms.BulkEditForm` with `nautobot.extras.forms.CustomFieldModelBulkEditFormMixin` and `nautobot.extras.forms.RelationshipModelBulkEditFormMixin`, and should be used where appropriate.
+* Similarly, for bulk-edit forms, `nautobot.extras.forms.NautobotBulkEditForm` combines `nautobot.utilities.forms.BulkEditForm` with `nautobot.extras.forms.CustomFieldBulkEditFormMixin`, `nautobot.extras.forms.RelationshipModelBulkEditFormMixin` and `nautobot.extras.forms.NoteModelBulkEditFormMixin`, and should be used where appropriate.
```
Comment (glennmatthews) [inline nautobot/docs/development/style-guide.md:66]:
```suggestion
* Similarly, for bulk-edit forms, `nautobot.extras.forms.NautobotBulkEditForm` combines `nautobot.utilities.forms.BulkEditForm` and `nautobot.utilities.forms.BootstrapMixin` with `nautobot.extras.forms.CustomFieldModelBulkEditFormMixin`, `nautobot.extras.forms.RelationshipModelBulkEditFormMixin` and `nautobot.extras.forms.NoteModelBulkEditFormMixin`, and should be used where appropriate.
```

File: nautobot/docs/development/style-guide.md [hunk 9]
```diff
@@ -63,7 +63,7 @@ New dependencies can be added to the project via the `poetry add` command. This
```
Comment (glennmatthews) [inline nautobot/docs/development/style-guide.md:63]: Would you mind also updating line 62 above:
- `CustomFieldModelForm` -> `CustomFieldModelFormMixin`
- `RelationshipModelForm` -> `RelationshipModelFormMixin`

File: nautobot/extras/forms/base.py [hunk 10]
```diff
@@ -21,10 +23,10 @@
 #
 
 
-class NautobotModelForm(BootstrapMixin, CustomFieldModelFormMixin, RelationshipModelFormMixin):
+class NautobotModelForm(BootstrapMixin, CustomFieldModelFormMixin, RelationshipModelFormMixin, NoteModelFormMixin):
     """
     This class exists to combine common functionality and is used to inherit from throughout the
-    codebase where all three of BootstrapMixin, CustomFieldModelFormMixin and RelationshipModelFormMixin are
+    codebase where all of BootstrapMixin, CustomFieldModelForm, RelationshipModelForm and NoteModelForm are
     needed.
```
Comment (glennmatthews) [inline nautobot/extras/forms/base.py:30]:
```suggestion
    codebase where all of BootstrapMixin, CustomFieldModelFormMixin, RelationshipModelFormMixin, and
    NoteModelFormMixin are needed.
```

File: nautobot/extras/forms/mixins.py [hunk 11]
```diff
@@ -137,6 +141,38 @@ def __init__(self, *args, **kwargs):
...
+    def save_note(self, *, instance, user):
+        if "object_note" in self.cleaned_data and self.cleaned_data.get("object_note"):
+            value = self.cleaned_data.get("object_note")
+            note = Note.objects.create(
+                note=value,
+                assigned_object_type=self.obj_type,
+                assigned_object_id=instance.pk,
+                user=user,
+            )
+            logger.debug("Created %s", note)
```
Comment (glennmatthews) [inline nautobot/extras/forms/mixins.py:158]: Perhaps avoid inadvertently creating whitespace-only Notes as well?

```suggestion
        value = self.cleaned_data.get("object_note", "").strip()
        if value:
            note = Note.objects.create(
                note=value,
                assigned_object_type=self.obj_type,
                assigned_object_id=instance.pk,
                user=user,
            )
            logger.debug("Created %s", note)
```

File: nautobot/docs/development/style-guide.md (hunk 9 above)
Comment (smk4664) [reply nautobot/docs/development/style-guide.md:63]: Sure

**Review by glennmatthews — APPROVED** [review]

**Pull request merged**